sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -0,0 +1,274 @@
+      None
+    } else {
+      // There must be only one plan at the last level, which contains all items.
+      assert(plansLastLevel.size == 1 && plansLastLevel.head._1.size == items.length)
+      Some(plansLastLevel.head._2.plan)
+    }
+  }
+
+  /** Find all possible plans in one level, based on previous levels. */
+  private def searchForLevel(level: Int): Unit = {
+    val foundPlansCurLevel = foundPlans(level)
+    var k = 1
+    var continue = true
+    while (continue) {
+      val otherLevel = level - k
+      if (k > otherLevel) {
+        // We can break from here, because when building a join from A and B, both A J B and B J A
+        // are handled.
+        continue = false
+      } else {
+        val joinPlansLevelK = foundPlans(k).values.toSeq
+        for (i <- joinPlansLevelK.indices) {
+          val curJoinPlan = joinPlansLevelK(i)
+
+          val joinPlansOtherLevel = if (k == otherLevel) {
+            // Both sides of a join are at the same level, no need to repeat for previous ones.
+            joinPlansLevelK.drop(i)
+          } else {
+            foundPlans(otherLevel).values.toSeq
+          }
+
+          joinPlansOtherLevel.foreach { otherJoinPlan =>
+            // Should not join two overlapping item sets.
+            if (curJoinPlan.itemIds.intersect(otherJoinPlan.itemIds).isEmpty) {
+              val joinPlan = buildJoin(curJoinPlan, otherJoinPlan)
+              if (joinPlan.nonEmpty) {
+                // Check if it's the first plan for the item set, or it's a better plan than
+                // the existing one due to lower cost.
+                val existingPlan = foundPlansCurLevel.get(joinPlan.get.itemIds)
+                if (existingPlan.isEmpty || joinPlan.get.cost < existingPlan.get.cost) {
+                  foundPlansCurLevel.update(joinPlan.get.itemIds, joinPlan.get)
+                }
+              }
+            }
+          }
+        }
+
+        k += 1
+      }
+    }
+  }
+
+  /** Build a new join node. */
+  private def buildJoin(curJoinPlan: JoinPlan, otherJoinPlan: JoinPlan): Option[JoinPlan] = {
+    // Check if these two nodes are inner joinable. We consider cartesian product very
+    // costly, thus exclude such plans. This also helps us to reduce the search space.
+    val curPlan = curJoinPlan.plan
+    val otherPlan = otherJoinPlan.plan
+    val joinCond = conditions
+      .filterNot(l => canEvaluate(l, curPlan))
+      .filterNot(r => canEvaluate(r, otherPlan))
+      .filter(e => e.references.subsetOf(curPlan.outputSet ++ otherPlan.outputSet))
+
+    if (joinCond.nonEmpty) {
+      val curPlanStats = curPlan.stats(conf)
+      val otherPlanStats = otherPlan.stats(conf)
+      if (curPlanStats.rowCount.nonEmpty && otherPlanStats.rowCount.nonEmpty) {
+        // Now both curPlan and otherPlan become intermediate joins, so the cost of the
+        // new join should also include their costs.
+        val cost = curJoinPlan.cost + otherJoinPlan.cost +
```

Comment:
But that means you are counting things twice right? Calling `LogicalPlan.stats(conf)` recursively determines costs for a `LogicalPlan`. The `JoinPlan` contains costs for the children of the join or 0 if there is no join. So if you combine those things seem to the duplicated, for example:
```
CURRENT = (TABLE A) JOIN (TABLE B) => Cost = c_A + c_B
OTHER = (TABLE C)
JOIN = ((TABLE A) JOIN (TABLE B) JOIN (TABLE C))
```
This given the following cost structure (`J` is the join statistics estimation function):
```
Costs(CURRENT) =  c_A + c_B
Costs(OTHER) = c_C
Costs(JOIN) = Costs(Current) + Cost(Other) + Cost((TABLE A) JOIN (TABLE B)) + Cost((TABLE C))
         = c_A + c_B + J(c_A, c_B) + c_C

```